bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/LastAddForcedCursor.java

```diff
@@ -0,0 +1,145 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.bookie;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import org.apache.bookkeeper.client.api.WriteAdvHandle;
+import org.apache.bookkeeper.proto.BookieProtocol;
+
+/**
+ * Handle the LastAddForced entryId. As entries do not arrive to the Journal in a known order we have to handle
+ * increments of this pointer. This is how it works: we keep a value for the current {@link #lastAddForced} entryId and
+ * a list of continuous ranges of ids which are known to have been persisted but are 'after' {@link #lastAddForced}.
+ * <p>
+ * Examples:
+ * <ul>
+ * <li>0, 1, 2, 3 we have maxAddForced = 3 (all entries from 0 to 3 are stored)</li>
+ * <li>0, 1, 3, 5 we have maxAddForced = 1 (all entries from 0 to 1 are stored, 2 and 4 are not stored)</li>
+ * <li>1, 2, 3 we have maxAddForced = -1 (not available, as 0 is missing, so there is an gap)</li>
+ * <li>&lt;empty sequence&gt; we have maxAddForced = -1 (not available)</li>
+ * </ul>
+ * </p>
+ * <p>
+ * When an gap is filled, like when we have a sequence of 0, 1, 2, 4, 5, 6, 8 and we receive a 3, we can advance
+ * {@link #lastAddForced} until we find a continuos sequence of ids, so we will get to 6.
+ * </p>
+ * <p>
+ * This cursor is kept in memory and in order to handle the case of Bookie restart we are reconstructing this
+ * {@link #lastAddForced} value from {@link LedgerStorage} just be asking for the maximum written entryId (with a scan)
```

Comment:
How long you keep the cursor in memory?

Reply:
We have a Guava cache. If there is no activity on a ledger or there are too many concurrent ledgers some cursor will be dropped and reconstructed if needed (by a force ledger call). We must make size and other  knobs of this cache tunable by configuration file

Reply:
We have a Guava cache. If there is no activity on a ledger or there are too many concurrent ledgers some cursor will be dropped and reconstructed if needed (by a force ledger call). We must make size and other  knobs of this cache tunable by configuration file

Reply:
Having a 'close' rpc will ease cache eviction